python/pyarrow/parquet/core.py

```diff
@@ -3424,16 +3426,20 @@ def write_metadata(schema, where, metadata_collector=None, **kwargs):
     ...     table.schema, 'dataset_metadata/_metadata',
     ...     metadata_collector=metadata_collector)
     """
-    writer = ParquetWriter(where, schema, **kwargs)
+    writer = ParquetWriter(where, schema, filesystem, **kwargs)
     writer.close()
 
     if metadata_collector is not None:
         # ParquetWriter doesn't expose the metadata until it's written. Write
         # it and read it again.
-        metadata = read_metadata(where)
+        metadata = read_metadata(where, filesystem=filesystem)
         for m in metadata_collector:
             metadata.append_row_groups(m)
-        metadata.write_metadata_file(where)
+        if filesystem is not None and not hasattr(where, "write"):
```

Comment:
```suggestion
        if filesystem is not None:
```

I would think we already ran into an error with `ParquetWriter` / `read_metadata` if the user would pass a filesystem in combination with a file-like object.

Reply:
But so actually, `read_metadata` does this by calling `filesystem, where = _resolve_filesystem_and_path(where, filesystem)`

We should probably do that here as well, so that the capabilities are consistent. For example that also supports writing to a URI string (you can test this with a local filesystem using `Path.as_uri()` to get a files:// URI)

Reply:
That's a lot better. :+1: